src/proto_nd_flow/reco/charge/raw_event_builder.py

```diff
@@ -97,6 +100,47 @@ def cross_rank_set_attrs(self, *attrs):
                 d = dict([(attr, getattr(self, attr)) for attr in attrs])
                 comm.send(d, dest=rank + 1)
 
+    @staticmethod
+    def unroll_timestamps(packets: np.ndarray) -> np.ndarray:
+        '''
+            Calculates "unrolled" timestamps for an array of packets. The
+            unrolled timestamps increase monotonically, rather than rolling over
+            every ~second. Each SYNC packet introduces an additional cumulative
+            offset (of ~1E7) that gets added to each subsequent raw timestamp,
+            giving the unrolled timestamps. Each IO group is independent.
+        '''
+        offsets = np.zeros((len(packets),), dtype='i8')
+        for io_group in np.unique(packets['io_group']):
+            mask = packets['io_group'] == io_group
+            sync_mask = (mask &
+                         (packets['packet_type'] == 6) &
+                         (packets['trigger_type'] == 83))
+            sync_ts = np.zeros_like(offsets)
+            # Replace 0 with ~1E7 at each SYNC; ~2E7 if PACMAN missed prev SYNC
+            sync_ts[sync_mask] = packets[sync_mask]['timestamp']
```

Comment:
We discussed this (using the timestamp of the PPS instead of the idealized 1E7) in-person, but I'm having some second thoughts...

Doing it this way is basically removing the procedure we rely on to keep the pacman clocks in sync. Over time these will drift apart and could cause issues when comparing timestamps across different io_groups. At the very least, this type of rollover accounting should happen after we calibrate the timestamps to account for the slightly different clock frequencies.

For now, I would suggest here we round `packets[sync_mask]['timestamp']` to the nearest `self.rollover_ticks`.

Reply:
Suggestion implemented.